### ISSUE TITLE
Support Logitech receiver with USB product ID 0xC542

### ISSUE THREAD

**Issue by nurietzsche2021** [pwr-Solaar/Solaar#1350]
- Solaar version: `1.1.0`
- Distribution: `Debian/Ubuntu`
- Kernel version : `Linux 5.13.0-21-generic x86_64 GNU/Linux`
- Output of `solaar show`:

<details>

```
Unifying Receiver
  Device path  : /dev/hidraw1
  USB id       : 046d:C52B
  Serial       : C826992E
    Firmware   : 24.11.B0036
    Bootloader : 02.09
    Other      : AA.AC
  Has 1 paired device(s) out of a maximum of 6.
  Notifications: wireless, software present (0x000900)
  Device activity counters: 1=226

  1: Wireless Touch Keyboard K400 Plus
     Device path  : /dev/hidraw2
     WPID         : 404D
     Codename     : K400 Plus
     Kind         : keyboard
     Protocol     : HID++ 4.1
     Polling rate : 8 ms (125Hz)
     Serial number: F3BB1CA8
     Model ID:      404D00000000
     Unit ID:       7E823222
        Bootloader: BOT 22.02.B0002
          Firmware: RQK 63.02.B0016
             Other: 
     The power switch is located on the top edge.
     Supports 24 HID++ 2.0 features:
         0: ROOT                   {0000}   
         1: FEATURE SET            {0001}   
         2: DEVICE FW VERSION      {0003}   
            Firmware: Bootloader BOT 22.02.B0002 404D6217AD40
            Firmware: Firmware RQK 63.02.B0016 404D6217AD40
            Firmware: Other   
            Unit ID: 7E823222  Model ID: 404D00000000  Transport IDs: {'wpid': '404D'}
         3: DEVICE NAME            {0005}   
            Name: Wireless Touch Keyboard K400 Plus
            Kind: keyboard
         4: WIRELESS DEVICE STATUS {1D4B}   
         5: RESET                  {0020}   
         6: BATTERY STATUS         {1000}   
            Battery: 20%, discharging, next level 5%.
         7: REPROG CONTROLS V4     {1B04}   
         8: SWAP BUTTON CANCEL     {2005}   
         9: NEW FN INVERSION       {40A2}   
            Fn-swap: enabled
            Fn-swap default: enabled
            Swap Fx function (saved): True
            Swap Fx function        : True
        10: ENCRYPTION             {4100}   
        11: KEYBOARD DISABLE KEYS  {4521}   
            Disable keys (saved): {'1': False, '16': False, '4': False, '8': False}
            Disable keys        : {'1': False, '4': False, '8': False, '16': False}
        12: TOUCHPAD RAW XY        {6100}   
        13: GESTURE 2              {6501}   
            Gestures (saved): {'1': True, '10': True, '2': True, '30': True, '34': False, '42': True, '43': True, '45': False, '64': False, '65': False, '67': False, '84': True}
            Gestures        : {'1': True, '2': True, '30': True, '10': True, '45': False, '42': True, '43': True, '64': False, '65': False, '67': False, '84': True, '34': False}
            Gesture params (saved): {'4': {'scale': 256}}
            Gesture params        : {'4': {'scale': 256}}
        14: DFUCONTROL UNSIGNED    {00C1}   
        15: unknown:1811           {1811}   internal, hidden
        16: unknown:1830           {1830}   internal, hidden
        17: unknown:1890           {1890}   internal, hidden
        18: unknown:1DF3           {1DF3}   internal, hidden
        19: unknown:1E00           {1E00}   hidden
        20: unknown:1EB0           {1EB0}   internal, hidden
        21: unknown:1861           {1861}   internal, hidden
        22: unknown:18B0           {18B0}   internal, hidden
        23: unknown:1F11           {1F11}   internal, hidden
     Has 20 reprogrammable keys:
         0: Second Left Click         , default: Left Click                  => Left Click                
             mse, divertable, pos:0, group:0, group mask:empty
             reporting: default
         1: Back                      , default: Mouse Back Button           => Mouse Back Button         
             is FN, FN sensitive, reprogrammable, divertable, pos:1, group:0, group mask:empty
             reporting: default
         2: My Home                   , default: HomePage                    => HomePage                  
             is FN, FN sensitive, reprogrammable, divertable, pos:2, group:0, group mask:empty
             reporting: default
         3: Application Switcher      , default: Application Switcher        => Application Switcher      
             is FN, FN sensitive, reprogrammable, divertable, pos:3, group:0, group mask:empty
             reporting: default
         4: Contextual Menu           , default: Right Click                 => Right Click               
             is FN, FN sensitive, reprogrammable, divertable, pos:4, group:0, group mask:empty
             reporting: default
         5: Search                    , default: Search Files                => Search Files              
             is FN, FN sensitive, reprogrammable, divertable, pos:5, group:0, group mask:empty
             reporting: default
         6: Show Desktop              , default: Show Desktop                => Show Desktop              
             is FN, FN sensitive, reprogrammable, divertable, pos:6, group:0, group mask:empty
             reporting: default
         7: Maximize Window           , default: Maximize Window             => Maximize Window           
             is FN, FN sensitive, reprogrammable, divertable, pos:7, group:0, group mask:empty
             reporting: default
         8: Switch Screen             , default: Switch Presentation/Switch Screen => Switch Presentation/Switch Screen
             is FN, FN sensitive, reprogrammable, divertable, pos:8, group:0, group mask:empty
             reporting: default
         9: Media Player              , default: Music                       => Music                     
             is FN, FN sensitive, reprogrammable, divertable, pos:9, group:0, group mask:empty
             reporting: default
        10: Previous                  , default: Previous                    => Previous                  
             is FN, FN sensitive, divertable, pos:10, group:0, group mask:empty
             reporting: default
        11: Play/Pause                , default: Play/Pause                  => Play/Pause                
             is FN, FN sensitive, divertable, pos:11, group:0, group mask:empty
             reporting: default
        12: Next                      , default: Next                        => Next                      
             is FN, FN sensitive, divertable, pos:12, group:0, group mask:empty
             reporting: default
        13: Fn Left Click             , default: ShowUI                      => ShowUI                    
             divertable, pos:0, group:0, group mask:empty
             reporting: default
        14: Left Button               , default: Left Click                  => Left Click                
             mse, pos:0, group:0, group mask:empty
             reporting: default
        15: Right Button              , default: Right Click                 => Right Click               
             mse, pos:0, group:0, group mask:empty
             reporting: default
        16: Mute                      , default: Mute                        => Mute                      
             nonstandard, divertable, pos:0, group:0, group mask:empty
             reporting: default
        17: Volume Up                 , default: Volume Up                   => Volume Up                 
             nonstandard, divertable, pos:0, group:0, group mask:empty
             reporting: default
        18: Volume Down               , default: Volume Down                 => Volume Down               
             nonstandard, divertable, pos:0, group:0, group mask:empty
             reporting: default
        19: Sleep                     , default: Sleep                       => Sleep                     
             reprogrammable, divertable, pos:0, group:0, group mask:empty
             reporting: default
     Has 17 gesture(s), 1 param(s) and 5 spec(s):
        Track1Finger               Enabled (None): None
        TapGestures                Enabled (None): None
        Tap1Finger                 Enabled (   0): True
        Tap2Finger                 Enabled (   1): True
        TapDrag1Finger             Enabled (   2): True
        DoubleTap1Finger           Enabled (   3): True
        Scroll2FingerStateless     Enabled (None): None
        NaturalScrolling           Enabled (   4): False
        Scroll2FingerHoriz         Enabled (   5): True
        Scroll2FingerVert          Enabled (   6): True
        LeftEdgeSwipe1Finger2      Enabled (   7): False
        RightEdgeSwipe1Finger2     Enabled (   8): False
        TopEdgeSwipe1Finger2       Enabled (   9): False
        Zoom2FingerStateless       Enabled (  10): True
        FnClickGestureSuppression  Enabled (  11): False
        TwoFingersPresent          Enabled (None): None
        DeviceSpecificRawData      Enabled (None): None
        ScaleFactor                Value   (   0): 256 [Default: 256]
        period unit                Spec    (   3): 2048
        DVI field width            Spec    (   1): 8
        field widths               Spec    (   2): 8
        multiplier                 Spec    (   5): None
        resolution                 Spec    (   4): 4
     Battery: 20%, discharging, next level 5%.

```
</details>

**Describe the bug**
I've read a few issues about inability to pair mouse M185 but I haven't found my solution yet. I've cloned the git repository in my Ubuntu 21.10 and I'm using the Solaar app to set my K400+ wireless keyboard, which paired automatically. However, when it comes to my Wireless Mouse M185, it works on my PC (indeed battery low warning appears when plugging USB) but device doesn't appear on apps' interface nor Solaar/bin solaar show (just the keyboard). I am also unable to pair neither with app nor the Solaar/bin solar pair command on terminal. It seems like my mouse is somehow invisible to Solaar even when I checked this repository to confirm that M185 is supposedly compatible with this engine.

Thanks for your support in advantance, I'm a training software developer and I'm still confused when messing with programmes files.

**Comment by pfps** [comment]
Devices don't pair automatically, so it is likely that your K400+ was paired with your Unifying receiver by Logitech before it was sold.  
M185 mice appear to come in several versions.  Only those that have a Unifying logo will be able to pair with a Unifying receiver.  If you don't have such an M185 then you cannot pair it with your Unifying receiver.
You should be able to use your M185 by using the receiver it came with.

**Comment by nurietzsche2021** [comment]
So I've found that my mouse doesn't have the unifying logo so I cannot pair using the Unifying receiver. I'm really lost on how to pair M185 with another receiver ... Could you guide me?

**Comment by pfps** [comment]
Your M185 came paired with some receiver.   You can only pair your M185 with receivers that are functionally the same.  But it may be that there is no way to make this determination without actually trying to pair.  So try to find a receiver that looks the same, and has the same USB ID, then plug that receiver into your Linux computer and see whether you can actually do the pairing.

The USB ID of a receiver can be shown using `lsusb`.   My Unifying receiver shows up as 
```
Bus 003 Device 016: ID 046d:c52b Logitech, Inc. Unifying Receiver
```

**Comment by nurietzsche2021** [comment]
As I was saying in the first message, my mouse does work in my computer. I only wanted some software to configure the mouse functioning (in my case regarding the scroll speed). I gave up on trying to find Logitechs' personalized applications to access my mouse settings and I decided to use general ubuntu configurations of mouse peripherals. For example, I used imwheel tool to stablish correct settings of mouse. I can leave this thread since I've seen that Solaar only works for Unifying receivers and there's nothing simillar for the rest of Logitech's devices.

**Comment by pfps** [comment]
If you have the receiver for the M185 mouse connected to your Linux box, you should be able to see the receiver and mouse in Solaar and control some aspects of the mouse.   If not, then please provide the output of `lsusb` so that the receiver can be added to the receivers that Solaar supports.

**Comment by nurietzsche2021** [comment]
As I was saying, my mouse doesn't appear on Solaar so I think receiver might not be available to use with this app. I'm going to leave here the output of my `lsusb`:
`Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 004: ID 5986:9102 Acer, Inc BisonCam,NB Pro
Bus 001 Device 003: ID 046d:c542 Logitech, Inc. Wireless Receiver
Bus 001 Device 002: ID 046d:c52b Logitech, Inc. Unifying Receiver
Bus 001 Device 005: ID 8087:0029 Intel Corp. AX200 Bluetooth
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub`

Thanks for your assistance

**Comment by pfps** [comment]
The receiver with ID 046d:c542 is not known to Solaar and needs to be added.  I'll create a patch for this.

**Comment by pfps** [comment]
PR #1354 should add support for your receiver 0xC542 and manipulate the mouse paired with it.

To download and use Solar from its GitHub repository
```
git clone https://github.com/pwr-Solaar/Solaar.git
cd Solaar
```
Run Solaar as bin/solaar from this directory.


To run PR #1354, first clone Solaar if you have not already done so and cd to the clone directory.  The first time you download the pull request, fetch it into a new branch and checkout that branch, as in:
```
git fetch origin pull/1354/head:pull_1354
git checkout pull_1354
```
To download a new version of the pull request, fetch it and then set your pull branch
to the new fetch, as in:
```
git checkout pull_1354
git fetch origin pull/1354/head
git reset --hard FETCH_HEAD
```

**Comment by nurietzsche2021** [comment]
I've already followed your steps and my mouse won't appear on Solaar's panel idk why. Only my K400+ handled by Unifying Receiver appears....

**Comment by pfps** [comment]
Only one instance of Solaar will run at one time so you have to quit out of any running instance of Solaar before starting a new one.
If this is not the problem, then please provide the output of `ls -l /dev/hidraw*` and `bin/solaar -dd show`.

**Comment by nurietzsche2021** [comment]
bin/solaar -dd show saturates my terminal and so much info appears that I cannot reach everything by scrolling up. So maybe it's enough if I save the output in a file and attach it?

**Comment by pfps** [comment]
Yes, you could redirect the output, both standard and error, to a file and attach that.

**Comment by pfps** [comment]
Also useful would be the output of `xinput`

**Comment by nurietzsche2021** [comment]
Output of `xinput`: `WARNING: running xinput against an Xwayland server. See the xinput man page for details.
⎡ Virtual core pointer                    	id=2	[master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer              	id=4	[slave  pointer  (2)]
⎜   ↳ xwayland-pointer:18                     	id=6	[slave  pointer  (2)]
⎜   ↳ xwayland-relative-pointer:18            	id=7	[slave  pointer  (2)]
⎣ Virtual core keyboard                   	id=3	[master keyboard (2)]
    ↳ Virtual core XTEST keyboard             	id=5	[slave  keyboard (3)]
    ↳ xwayland-keyboard:18                    	id=8	[slave  keyboard (3)]`

**Comment by nurietzsche2021** [comment]
I ran `bin/solaar -dd show > solaar_show`and this is the file that I'm attaching
[solaar_show.txt](https://github.com/pwr-Solaar/Solaar/files/7625943/solaar_show.txt)

**Comment by pfps** [comment]
You are running Xwayland, which doesn't provide good output.   Is there a way to list the input devices in Wayland?  I want to check whether both your devices show up there.

**Comment by pfps** [comment]
You need to redirect both the standard output and the error output to the file so that the debugging output goes there too.  In bash this is done using `&>`, I believe.

**Comment by pfps** [comment]
The hidraw output looks a bit unusual as well.  Can you copy this shell script and run it?  This should show which hidraw nodes correspond to your Logitech devices.

```
#!/bin/bash

FILES=/dev/hidraw*
for f in $FILES
do
  FILE=${f##*/}
  DEVICE="$(cat /sys/class/hidraw/${FILE}/device/uevent | grep HID_NAME | cut -d '=' -f2)"
  ID="$(cat /sys/class/hidraw/${FILE}/device/uevent | grep HID_ID | cut -d '=' -f2)"
  printf "%s %s %s\n" $FILE "$ID" "$DEVICE" 
done
```

**Comment by nurietzsche2021** [comment]
> You are running Xwayland, which doesn't provide good output. Is there a way to list the input devices in Wayland? I want to check whether both your devices show up there.

I'm using the libinput list-devices that I found in some page that's supposed to be the equivalent of xinput for Wayland.
This is the output:
[list_devices.txt](https://github.com/pwr-Solaar/Solaar/files/7626222/list_devices.txt)

**Comment by nurietzsche2021** [comment]
> The hidraw output looks a bit unusual as well. Can you copy this shell script and run it? This should show which hidraw nodes correspond to your Logitech devices.
> 
> ```
> #!/bin/bash
> 
> FILES=/dev/hidraw*
> for f in $FILES
> do
>   FILE=${f##*/}
>   DEVICE="$(cat /sys/class/hidraw/${FILE}/device/uevent | grep HID_NAME | cut -d '=' -f2)"
>   ID="$(cat /sys/class/hidraw/${FILE}/device/uevent | grep HID_ID | cut -d '=' -f2)"
>   printf "%s %s %s\n" $FILE "$ID" "$DEVICE" 
> done
> ```

Output:
`hidraw0 0018:000004F3:00003162 ELAN0412:01 04F3:3162
hidraw1 0003:0000046D:0000C52B Logitech USB Receiver
hidraw2 0003:0000046D:0000404D Logitech K400 Plus
hidraw4 0003:0000046D:0000C542 Logitech Wireless Receiver `

**Comment by pfps** [comment]
From the list-devices output it appears that your two receivers are supported at different levels.  The keyboard is show with its product name, but the mouse is just a generic mouse.  This is not surprising, as new hardware takes a while to get full support in Linux.  It is unclear whether this affects 
Solaar.
The output from the shell script also shows the difference, with the Unifying receiver being listed and the keyboard listed as well but the mouse connected to the Nano receiver not being listed.  Again, this may or may not affect Solaar.

The debug output from Solaar shows a problem.  The HID interface for the Nano receiver is interface 0 but Solaar is expecting it to be interface 1.

```
15:00:22,033    DEBUG [MainThread] hidapi.udev: Found device BID 0003 VID 0000046D PID 0000C542 INTERFACE 0 FILTER 1
```
There are two possibilities here.  Either your kernel does not support this receiver well enough for Solaar to work with it or the receiver is different from other Logitech receivers.  To start figuring out which is the case, please post output of `lsusb -v -d 046d:c542`.  This will show the interfaces that the receiver has.

**Comment by nurietzsche2021** [comment]
Here's the output: 
`Bus 001 Device 003: ID 046d:c542 Logitech, Inc. Wireless Receiver
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               1.10
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0         8
  idVendor           0x046d Logitech, Inc.
  idProduct          0xc542 
  bcdDevice            3.02
  iManufacturer           1 Logitech
  iProduct                2 Wireless Receiver
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0022
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower               50mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode           33 US
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      90
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               4`

**Comment by pfps** [comment]
That's interesting - only one Interface Descriptor, so it is likely that that interface does provide what Solaar needs.  I'll update the PR to allow for that interface.

I figured out how to show the report descriptors on many Linux machines.  Could you try running this script as root and post the output?

```
#!/bin/bash
lsusb -v -d 046D:C542
FILES=/sys/kernel/debug/hid/0003:046D:C542.*
for f in $FILES
do
  echo ''
  report=`cat $f/rdesc | head -n1`
  let lgth=${#report}/3
  echo REPORT $f ${lgth}
  echo ${report}
done
```

**Comment by nurietzsche2021** [comment]
I ran the script using sudo and that's what I got:
`Bus 001 Device 003: ID 046d:c542 Logitech, Inc. Wireless Receiver
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               1.10
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0         8
  idVendor           0x046d Logitech, Inc.
  idProduct          0xc542 
  bcdDevice            3.02
  iManufacturer           1 Logitech
  iProduct                2 Wireless Receiver
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0022
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower               50mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode           33 US
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      90
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               4
can't get debug descriptor: Resource temporarily unavailable
Device Status:     0x0000
  (Bus Powered)

REPORT /sys/kernel/debug/hid/0003:046D:C542.0005 90
05 01 09 02 a1 01 85 01 09 01 a1 00 05 09 19 01 29 05 15 00 25 01 95 05 75 01 81 02 95 01 75 03 81 01 05 01 09 30 09 31 16 01 80 26 ff 7f 75 10 95 02 81 06 09 38 15 81 25 7f 75 08 95 01 81 06 c0 c0 05 01 09 00 a1 01 85 05 06 00 ff 09 01 15 81 25 7f 75 08 95 07 b1 02 c0`

**Comment by pfps** [comment]
Hmm.  It might be that this receiver is not compatible with Solaar as it doesn't appear to have the interface that Solaar uses.

But try out the revised version of the PR #1354 with `bin/solaar -dd show` as before.

**Comment by nurietzsche2021** [comment]
Okaym I updated my branch pull_1354 with the commands you indicated me and I've created a new file with bin/solaar -dd show results and errors. Here it is:
[solaar_show.txt](https://github.com/pwr-Solaar/Solaar/files/7627619/solaar_show.txt)

**Comment by pfps** [comment]
That's very strong evidence that your receiver will not work with Solaar, as it does not respond to any HID++ (the protocol Solaar uses) commands.  I'm going to document this state of affairs as all that can be done unless I get information from Logitech that the receiver does use HID++.
